Add blog post about K3s bootstrap (#396)
k3s-io/docs: blog/2025-03-25-K3s-initialization.md

---
title: "K3s initialization deep dive"
description: Explain k3s initialization steps
authors: manuelbuil # Fill out your info in the authors.yml file
hide_table_of_contents: true
---
K3s is a lightweight Kubernetes distribution which excels in its deployment speed and minimal resource footprint. In fact, a lot of our users love K3s because it offers an unparalleled initialization speed.

<!-- truncate -->
This blog post delves into the heart of K3s's efficiency: its initialization process. We'll embark on a journey through the steps that enable K3s to materialize a fully functional Kubernetes cluster so quickly. By examining K3s’s own logs, we'll unravel the meaning behind each step, providing you with a practical understanding of how K3s achieves its remarkable speed. This exploration not only illuminates the inner workings of K3s but also equips you with the knowledge to troubleshoot your deployments.

## The Embedded Powerhouse ⚙️⚡

> K3s leverages `go-bindata` to embed essential Linux userspace binaries and Helm charts directly into its executable. This eliminates external dependencies and streamlines the deployment process. Within the K3s binary, you'll find core components like `runc` and `containerd`, along with the k3s-root tarball (e.g. k3s-root-amd64.tar). This tarball contains all the userspace binaries necessary for K3s to function, reducing the reliance on the host OS. If the binary is already included in the host OS, it will have preference, unless we use the `--prefer-bundle-bin` flag.

Author: ah, interesting, thanks for the update Brad. Let me see how can I rewrite that without getting too much into detail

> K3s leverages `go-bindata` to embed essential Linux userspace binaries and Helm charts directly into its executable. This eliminates external dependencies and streamlines the deployment process. Within the K3s binary, you'll find core components like `runc` and `containerd`, along with the k3s-root tarball (e.g. k3s-root-amd64.tar). This tarball contains all the userspace binaries necessary for K3s to function, reducing the reliance on the host OS. If the binary is already included in the host OS, it will have preference, unless we use the `--prefer-bundle-bin` flag.

Author: I decided to not clarify what happens in the default case, as it would probably require to mention a few binaries and could change in the future. I wrote instead:
```
This tarball contains all the userspace binaries necessary for K3s to function, reducing the reliance on the host OS. If you would like all the K3s embedded binaries to take preference over the host OS binaries, you should use the `--prefer-bundle-bin` flag.
 ```

> K3s leverages `go-bindata` to embed essential Linux userspace binaries and manifests directly into its executable. This eliminates external dependencies and streamlines the deployment process. Within the K3s binary, you'll find core components like `runc` and `containerd`, along with the k3s-root tarball (e.g. k3s-root-amd64.tar). This tarball contains all the userspace binaries necessary for K3s to function, reducing the reliance on the host OS. If you would like all the K3s embedded binaries to take preference over the host OS binaries, you should use the `--prefer-bundle-bin` flag.

Reviewer: ```suggestion
K3s leverages `go-bindata` to embed essential Linux userspace binaries and manifests directly into its executable. This eliminates external dependencies and streamlines the deployment process. Within the K3s binary, you'll find core components like `runc` and `containerd`, along with the k3s-root tarball (e.g. k3s-root-amd64.tar). This tarball contains all the userspace binaries necessary for K3s to function, reducing the reliance on the host OS. If you would like all the K3s embedded binaries to take preference over the host OS binaries, you should use the `--prefer-bundled-bin` flag.
```

Author: oh no! Thanks!

```
Starting Lightweight Kubernetes...
```

And then:

```
/usr/bin/systemctl is-enabled --quiet nm-cloud-setup.service
```

This part is checking for a network manager utility which must be disabled as described in the [docs](https://docs.k3s.io/installation/requirements?_highlight=nm&_highlight=cloud&_highlight=setup.service&os=rhel#operating-systems). It configures some parts of the network stack, specifically the routing tables, which conflict with Kubernetes networking, and that is why we verify if it was correctly disabled.

The next message should look familiar

```
Acquiring lock file /var/lib/rancher/k3s/data/.lock
Preparing data dir /var/lib/rancher/k3s/data/f8e9b5e7d85085972f4a9ddfd539d4dcf887be2e380a55f415c93cac5516dad5
```

When this message is shown, the directory where K3s deploys the embedded binaries has already been created. At this point, K3s will extract the binaries. We use the lock to avoid concurrent modifications, preventing K3s embedded commands like `kubectl` or `ctr` from executing and disturbing the K3s initialization.

> The next block of logs point at the K3s version and the datastore. In this case, I am deploying the default K3s which means sqlite. For more information on the different datastores available check this [link](https://docs.k3s.io/datastore)

Reviewer: ```suggestion
The next block of logs point at the K3s version and the datastore. In this case, I am using the default datastore -which means kine with sqlite. For more information on the different datastores available check this [link](https://docs.k3s.io/datastore)
```

certificate CN=kube-apiserver signed by CN=k3s-server-ca@1742309831
generated self-signed CA certificate CN=k3s-request-header-ca@1742309831
certificate CN=system:auth-proxy signed by CN=k3s-request-header-ca@1742309831
generated self-signed CA certificate CN=etcd-server-ca@1742309831
certificate CN=etcd-client signed by CN=etcd-server-ca@1742309831
generated self-signed CA certificate CN=etcd-peer-ca@1742309831
certificate CN=etcd-peer signed by CN=etcd-peer-ca@1742309831
certificate CN=etcd-server signed by CN=etcd-server-ca@1742309831
certificate CN=k3s,O=k3s signed by CN=k3s-server-ca@1742309831
```

And then saves the bootstrap data in the bootstrap key. Again, this is not relevant for this example:

```
Saving cluster bootstrap data to datastore
```

After that, K3s starts the different Kubernetes components. These components are all run within the k3s process as goroutines, which are lightweight, concurrent functions in Go, allowing for efficient resource usage. This is another design decision taken to reduce boot time and reduce resource consumption.

```
Running kube-apiserver
Running kube-scheduler
Running kube-controller-manager
```

> Just for completeness, the charts are extracted in the directory `/var/lib/rancher/k3s/server/manifests/` when all the different Kubernetes components are running and K3s initialization is ready. The deploy controller watches this directory and applies all the manifests. This is how components like CoreDNS or Traefik eventually get installed

Reviewer: ```suggestion
Manifests for packaged components are extracted to `/var/lib/rancher/k3s/server/manifests/`. When all the different Kubernetes components are running and K3s initialization is ready, the deploy controller begins watching this directory and applies all the manifests. This is how components like CoreDNS or Traefik eventually get installed.
```

> Just for completeness, the charts are extracted in the directory `/var/lib/rancher/k3s/server/manifests/` when all the different Kubernetes components are running and K3s initialization is ready. The Helm Controller, which is a K3s embedded controller and its explanation is out of the scope for this post, will install any chart located in that directory. This is how components like CoreDNS or Traefik eventually get installed

Reviewer: Technically the helm controller does not interact with this directory. The deploy controller watches this directory and applies all the manifests. If the manifest contains a HelmChart or HelmChartConfig resource, the helm controller will pick that up once the resource has been created in the cluster. The helm controller does not at any point interact directly with files in this directory.

Author: You are 100% right, thanks!
